components/layout_2020/flow/root.rs

```diff
@@ -229,15 +200,165 @@ impl FragmentTreeRoot {
         ))
     }
 
-    pub fn bounding_box_of_border_boxes(&self) -> Rect<Au> {
-        let origin = Point2D::new(
-            Au::from_f32_px(self.bounding_box_of_border_boxes.origin.x.px()),
-            Au::from_f32_px(self.bounding_box_of_border_boxes.origin.y.px()),
-        );
-        let size = Size2D::new(
-            Au::from_f32_px(self.bounding_box_of_border_boxes.size.width.px()),
-            Au::from_f32_px(self.bounding_box_of_border_boxes.size.height.px()),
-        );
-        Rect::new(origin, size)
+    fn iterate_through_fragments(&self, iterator: &mut dyn FragmentIterator) {
```

Comment:
```suggestion
    fn find_fragment<T>(&self, iterator: impl FnMut(&Fragment, &PhysicalRect<Length>) -> Option<T>) -> Option<T> {
```

* Why `dyn Trait` over `impl Trait`? I tend to default monomorphized static dispatch because it gives more inlining opportunities to the optimizer, but if there’s a reason to prefer dynamic dispatch I’m not opposed
* Both impls of the `FragmentIterator` trait return `false` when they’ve found an appropriate fragment, and store some result in a field as a side effect. If we make that result be part of the return value instead (by replacing `bool` with `Option`) them those fields initialized to a zero/dummy value would not be needed.
* With the change above there would be no need to access fields after iteration as ended, so by replacing the custom trait with `FnMut` we could use closure syntax and not need to define explicit structs, which would be more concise.

What do you think?

Reply:
> Why dyn Trait over impl Trait? I tend to default monomorphized static dispatch because it gives more inlining opportunities to the optimizer, but if there’s a reason to prefer dynamic dispatch I’m not opposed

There's no explicit reason. I've modified this to use `impl Trait` static dispatch.

> Both impls of the FragmentIterator trait return false when they’ve found an appropriate fragment, and store some result in a field as a side effect. If we make that result be part of the return value instead (by replacing bool with Option) them those fields initialized to a zero/dummy value would not be needed.

One issue is that some of these iterators will need to accumulate values. In fact, I have just modified the getBoundingClientRect iterator to accumulate a single rectangle for every fragment that has the appropriate tag. I think this is what should be done here. Perhaps it's still possible to do this with two generic type parameters?

Reply:
With `FnMut` (as opposed to `Fn`) closures can mutate their environment / captures. So accumulating things across calls should not be a problem. Would you like me to try moving to closures?

Reply:
Oops. I think I misinterpreted your `FnMut` suggestion. I've pushed a new version of the branch that does away with the traits and uses mutable closures.